backends-velox/src/main/scala/org/apache/spark/sql/execution/BroadcastUtils.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution
+
+import io.glutenproject.columnarbatch.ColumnarBatches
+import io.glutenproject.memory.nmm.NativeMemoryManagers
+import io.glutenproject.vectorized.{ColumnarBatchSerializeResult, ColumnarBatchSerializerJniWrapper}
+
+import org.apache.spark.SparkContext
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.plans.physical.{BroadcastMode, BroadcastPartitioning, Partitioning}
+import org.apache.spark.sql.execution.joins.HashedRelation
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.util.TaskResources
+
+import scala.collection.JavaConverters._;
+
+// Utility methods to convert Vanilla broadcast relations from/to Velox broadcast relations.
+// FIXME: Add checking for broadcast mode.
+// FIXME: Truncate output with batch size.
+object BroadcastUtils {
+  def veloxToSparkUnsafe[F, T](
+      context: SparkContext,
+      mode: BroadcastMode,
+      from: Broadcast[F],
+      fn: Iterator[ColumnarBatch] => Iterator[InternalRow]): Broadcast[T] = {
+    // ColumnarBuildSideRelation to HashedRelation.
+    val fromBroadcast = from.asInstanceOf[Broadcast[ColumnarBuildSideRelation]]
+    val fromRelation = fromBroadcast.value.asReadOnlyCopy()
+    var rowCount: Long = 0
+    val toRelation = TaskResources.runUnsafe {
+      val rowIterator = fn(fromRelation.deserialized.flatMap {
+        cb =>
+          rowCount += cb.numRows()
+          Iterator(cb)
+      })
+      mode.transform(rowIterator, Some(rowCount))
+    }
+    // Rebroadcast Spark relation.
+    context.broadcast(toRelation).asInstanceOf[Broadcast[T]]
+  }
+
+  def sparkToVeloxUnsafe[F, T](
+      context: SparkContext,
+      mode: BroadcastMode,
+      schema: StructType,
+      from: Broadcast[F],
+      fn: Iterator[InternalRow] => Iterator[ColumnarBatch]): Broadcast[T] = {
+    // HashedRelation to ColumnarBuildSideRelation.
+    val fromBroadcast = from.asInstanceOf[Broadcast[HashedRelation]]
```

Comment:
Can you please add a check here for BroadcastMode? HashedRelation will only be present in case of `HashedRelationBroadcastMode` . It will be `Array[InternalRow]` in case of `IdentityBroadcastmode`

Reply:
just checked your comment  `FIXME: Add checking for broadcast mode.`

Reply:
Sure. Will add some checks for mode.